File: IndieLibX/Docs/ProceduralBuilding_Docs/CantorDust/CantorDust2D.py
#
# Copyright (c) 2018 Tommi Roenty   http://www.tommironty.fi/
# Licensed under The GNU Lesser General Public License, version 2.1:
#     http://opensource.org/licenses/LGPL-2.1
#

import math
import pygame
import sys

window=(640,512)
    
'''
cantordust
{
	instance "box"
	move [0, 120%, 0]
	divide x
		[width/3] cantordust
		[width/3] {}
		[width/3] cantordust
}

cantordust : width < 0.5 {}
'''

def CantorDust(screen, x, width, y,height):
	if  width < 0.5:
		return

	CantorDust(screen, x, 			width/3.0, y+height,height)
	CantorDust(screen, x+2.0*width/3.0, 	width/3.0, y+height,height)

	#start = [x, y]
	#end = [x+width, y ]
	pygame.draw.rect(screen, (0,0,0), (x,y,width,height), 1)

pygame.init()
pygame.display.set_caption('CantorDust2D')
screen = pygame.display.set_mode((window[0],window[1]))
screen.fill((255,255,255))

while (True):
	for event in pygame.event.get():
        	if event.type == pygame.QUIT:
             		pygame.quit()
			sys.exit()
	CantorDust(screen, 0.0, window[0], 0.0, 10.0)
	pygame.display.update()

File: IndieLibX_TinyXML/Samples/ProceduralBuilding/CantorDust2D.cpp
/*
  Copyright (c) 2018 Tommi Roenty   http://www.tommironty.fi/
  Licensed under The GNU Lesser General Public License, version 2.1:
      http://opensource.org/licenses/LGPL-2.1
*/
#include <Framework3/IRenderer.h>

void CantorDust(D3DXFROMWINEVECTOR2 aPosition, D3DXFROMWINEVECTOR2 aSize)
{
	if(aSize.x < 0.005)
		return;

	 D3DXFROMWINEVECTOR2 size(aSize.x/3.0,aSize.y);
	{D3DXFROMWINEVECTOR2 position(aPosition.x, aPosition.y+aSize.y);
	 CantorDust(position, size);}
	{D3DXFROMWINEVECTOR2 position(aPosition.x+2.0*aSize.x/3.0, aPosition.y+aSize.y);
	 CantorDust(position, size);}

	STX_Rect rekt;
	rekt.x = aPosition.x	*	IRenderer::GetRendererInstance()->GetViewportWidth();
	rekt.y = aPosition.y	*	IRenderer::GetRendererInstance()->GetViewportHeight();
	rekt.w = aSize.x	*	IRenderer::GetRendererInstance()->GetViewportWidth();
	rekt.h = aSize.y	*	IRenderer::GetRendererInstance()->GetViewportHeight();
	STX_RenderFillRect(rekt);
}

int init(const char* aTitle)
{
	return 0;
}

void render()
{
	float f=128.0f/256.0f;
	IRenderer::GetRendererInstance()->Clear(true, true, D3DXFROMWINEVECTOR4 (f, f, f, 1.0f));
	IRenderer::GetRendererInstance()->BeginScene();

	D3DXFROMWINEVECTOR2 aPosition(0.0f, 0.0f);
	D3DXFROMWINEVECTOR2 aSize(1.0f, 0.1f);
	CantorDust(aPosition, aSize);

	IRenderer::GetRendererInstance()->EndScene();
	IRenderer::GetRendererInstance()->Present( );
}

int ApplicationLogic()
{
	IRenderer* r=IRenderer::GetRendererInstance("CantorDust2D");
	IInput* in=STX_Service::GetInputInstance();
	init("");
	while (!in->OnKeyPress (STX_KEY_ESCAPE) && !in->Quit())
	{
		in->Update();
		render();
	}
	return 0;
}

